torch/_dynamo/variables/torch.py

```diff
@@ -855,12 +855,20 @@ def speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
                 # NB: we don't bother populating graphargs, as
                 # they won't actually get used by anything
 
-            output = f.call_function(tx, args, {})
+            try:
+                output = f.call_function(tx, args, {})
+            except Exception:
+                raise
 
             # Register output to graph
             # Modeled off of compile_and_call_fx_graph
             # TODO: support non single Tensor output
-            assert isinstance(output, TensorVariable)
+            if not isinstance(output, TensorVariable):
+                raise TypeError(
+                    "Expected branch out type to be a single tensor but got {}".format(
+                        str(output.python_type())
```

Comment:
nit: you don't need to str here, it will be implicitly str'ed by the format operation. Also, use an f-string, much nicer:
`f"Expected branch out type to be a single tensor but got {output.python_type()}"`

Reply:
It's not clear to me why this is a TypeError and not a UserError, under the logic for UserError? As in, not returning a Tensor from the subgraph is clearly a problem for the user to solve?

Also, if we're modifying this code, you may want to distinguish the case for not supporting multi-Tensor output (tuple/list return), since that is not a UserError.

Reply:
Yeah, it is a UserError. Keep it as an internal error here just because this function is not only used by cond(), so this exception is propagated up and is wrapped to be a UserError at the entry of cond().

Reply:
This seems fairly error prone in the long term? How are you going to get people to remember to use UserError or an internal error whenever they are writing errors? How are you going to make sure people consistently do the wrapping? (And how are you going to deal with people who are complaining that rewrapping exceptions blows up the stack trace?)